% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/process-get_mutation_status_of_read.R
\name{get_mutation_status_of_read}
\alias{get_mutation_status_of_read}
\title{Get mutation status of read}
\usage{
get_mutation_status_of_read(
  chr,
  pos,
  ref,
  alt,
  read_stats,
  read_index_at_pos,
  fasta_fafile = NULL,
  fasta_seq = NULL,
  cigar_free_indel_match = FALSE,
  n_match_base_before = 1,
  n_match_base_after = 1
)
}
\arguments{
\item{chr}{Character vector representing the chromosome of interest.}

\item{pos}{Numeric value representing the Genomic position of interest.}

\item{ref}{Character vector representing reference base(s).}

\item{alt}{Character vector representing alternative base(s).}

\item{read_stats}{A list of read-level statistics.}

\item{read_index_at_pos}{An integer representing the index of the nucleotide in sequence aligning with the position of interest.}

\item{fasta_fafile}{An open connection to an object of class FaFile}

\item{fasta_seq}{A list with the fasta sequence between two positions.}

\item{cigar_free_indel_match}{Boolean. If activated, the information from the CIGAR is disregarded when determining the
mutation status of a read for indel. Instead the mutation status is determined by comparing the sequence of the read
to the sequence of the wild-type reference and the mutated reference. Activating this option may lead to discordant
genotyping of reads compared to the information provided by the CIGAR for indels. On the other hand, when
activated, it may rescue mutated genotypes for indel that would be missed in cases where the representation of the
indel in the CIGAR does not match the norm of bcftools of the mutation being analyzed.}

\item{n_match_base_before}{Number of bases to be matched before the alt allele in the sequences comparison}

\item{n_match_base_after}{Number of bases to be matched after the last alt allele in the sequences comparison}
}
\description{
The algorithm to determine if a read supports the mutation of interest is quite intuitive but requires careful
considerations of insertions and deletions in repeated sequences to resolve ambiguities whenever possible. The core
idea is to apply the mutation on the reference sequence and compare the mutated reference sequence to the read
sequence starting from the position of the mutation of interest. The sequence of the read should be compared to the
sequence of the reference sequence and of the mutated reference sequence for just enough bases to assign mutational
status without ambiguity.
}
\keyword{internal}
