% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/prepro-read_bam.R
\name{read_bam}
\alias{read_bam}
\title{Extract and filter paired-end reads for a target locus from a BAM file}
\usage{
read_bam(
  bam,
  chr,
  pos,
  neg_offset_mate_search,
  pos_offset_mate_search,
  flag_bam_list
)
}
\arguments{
\item{bam}{Path to a BAM file.}

\item{chr}{Character. Chromosome of interest.}

\item{pos}{Integer. Genomic position of interest.}

\item{neg_offset_mate_search}{Integer. Use in read_bam.
Represents the number of nucleotides to extend upstream (negative direction)
from the position of interest when querying the BAM file with Rsamtools.
his extension ensures that paired reads are retrieved, even if only one mate
overlaps the queried position.}

\item{pos_offset_mate_search}{Integer. Use in read_bam.}

\item{flag_bam_list}{A named list of logicals for filtering reads based on their SAM flag
NA = Filter is ignored, TRUE = The read MUST have this flag, FALSE = The read MUST NOT have this flag.}
}
\value{
A dataframe containing the filtered SAM entries.
}
\description{
This function performs a targeted extraction of sequencing reads from a BAM file. It first fetches reads
within a specified genomic window around a variant of interest, then expands the selection to include the mates of
any reads covering the variant, ensuring complete fragments are retrieved for analysis.
}
\details{
The read retrieval and filtering process follows a multi-step pipeline:
\enumerate{
\item A genomic query region is defined around the target 'pos' using the 'neg_offset_mate_search' and 'pos_offset_mate_search' parameters.
\item 'Rsamtools::scanBam' is used to fetch all reads within this region that pass the filters specified by 'flag_bam_list'.
\item From this initial set, the function identifies the subset of reads that \emph{directly} cover the specific 'pos'.
\item The names ('QNAME') of the fragments corresponding to these covering reads are collected.
\item The function then selects \strong{all} reads from the initially fetched data that share these fragment names, thereby
retrieving the mates even if they did not directly cover the variant position.
\item A final filter is applied to retain only properly oriented pairs, where the read and its mate have opposite strand orientations.
}
}
\keyword{internal}
